File: lang/nodejs/src/fun/vac/data_structures/ArrayStack.js
"use strict"

exports.ArrayStack = class {

    constructor (physicalSize = 0) {
        this.data = null
        this.logicalSize = 0
        this.physicalSize = 64

        if (physicalSize > 1) {
            this.physicalSize = physicalSize
        }
        this.data = new Array(this.physicalSize)
    }

    size () {
        return this.logicalSize
    }

    isEmpty () {
        return this.logicalSize === 0
    }

    peek () {
        if (this.logicalSize === 0) {
            throw new Error("[PANIC - NoSuchElement]")
        }

        return this.data[this.logicalSize - 1]
    }

    push (element) {
        if (this.logicalSize === this.physicalSize) {
            let newCapacity = 64

            if (this.physicalSize >= 64) {
                newCapacity = this.physicalSize + (this.physicalSize >>> 1)
            }

            const temp = new Array(newCapacity)

            for (let i = 0, length = this.logicalSize; i < length; i++) {
                temp[i] = this.data[i]
            }

            this.data = temp
            this.physicalSize = newCapacity
        }

        this.data[this.logicalSize] = element

        this.logicalSize++
    }

    pop () {
        if (this.logicalSize === 0) {
            throw new Error("[PANIC - NoSuchElement]")
        }

        this.logicalSize--

        this.data[this.logicalSize] = null
    }

    capacity () {
        return this.physicalSize
    }

    shrink () {
        const temp = new Array(this.logicalSize)

        for (let i = 0, length = this.logicalSize; i < length; i++) {
            temp[i] = this.data[i]
        }

        this.data = temp
        this.physicalSize = this.logicalSize
    }
}

File: go/muse/data_structures/array_stack/array_stack.go
package array_stack

import "errors"

const DEFAULT_CAPACITY int = 64

type ArrayStack[T any] struct {
	data         []T
	logicalSize  int
	physicalSize int
}

func New[T any]() *ArrayStack[T] {
	return NewWithCapacity[T](0)
}

func NewWithCapacity[T any](physicalSize int) *ArrayStack[T] {
	stack := &ArrayStack[T]{data: nil, logicalSize: 0, physicalSize: DEFAULT_CAPACITY}

	if physicalSize > 1 {
		stack.physicalSize = physicalSize
	}
	stack.data = make([]T, stack.physicalSize)

	return stack
}

func (s *ArrayStack[T]) Size() int {
	return s.logicalSize
}

func (s *ArrayStack[T]) IsEmpty() bool {
	return s.logicalSize == 0
}

func (s *ArrayStack[T]) Peek() T {
	if s.logicalSize == 0 {
		panic(errors.New("[PANIC - NoSuchElement]"))
	}

	return s.data[s.logicalSize-1]
}

func (s *ArrayStack[T]) Push(element T) {
	if s.logicalSize == s.physicalSize {
		newCapacity := DEFAULT_CAPACITY

		if s.physicalSize >= DEFAULT_CAPACITY {
			newCapacity = s.physicalSize + (s.physicalSize >> 1)
		}

		temp := make([]T, newCapacity)

		for i := range s.logicalSize {
			temp[i] = s.data[i]
		}

		s.data = temp
		s.physicalSize = newCapacity
	}

	s.data[s.logicalSize] = element

	s.logicalSize++
}

func (s *ArrayStack[T]) Pop() {
	if s.logicalSize == 0 {
		panic(errors.New("[PANIC - NoSuchElement]"))
	}

	s.logicalSize--

	s.data[s.logicalSize] = *new(T)
}

func (s *ArrayStack[T]) Capacity() int {
	return s.physicalSize
}

func (s *ArrayStack[T]) Shrink() {
	temp := make([]T, s.logicalSize)

	for i := range s.logicalSize {
		temp[i] = s.data[i]
	}

	s.data = temp
	s.physicalSize = s.logicalSize
}
